agent/session.go

```diff
@@ -54,6 +56,7 @@ func newSession(ctx context.Context, agent *Agent, delay time.Duration, sessionI
 		subscriptions: make(chan *api.SubscriptionMessage),
 		registered:    make(chan struct{}),
 		closed:        make(chan struct{}),
+		cancel:        sessionCancel,
```

Comment:
I think either would work.  If the session fails or times out, the session needs to be closed and restarted anyway, so I think the same session + cancellation can be applicable.  I'm happy to go either way.